### DISCUSSION TITLE
How to override LinkNode?

### DISCUSSION THREAD

**Discussion by kazukinagata** [facebook/lexical#3559] [Q&A]
I wanted to override LinkNode's functionality, so I created and registered the `CustomLinkNode` following this [document](https://lexical.dev/docs/concepts/node-replacement).

However, when I actually insert the link, the log does not output "hello custom link", and when I check the TreeView, it seems that "link" is used instead of "custom-link".

Here is a simple [demo](https://codesandbox.io/s/nifty-tdd-hif7gw?file=/src/nodes/CustomLinkPlugin.js) .

I am new to Lexical. Is there anything I am missing?

```javascript
import { LinkNode } from "@lexical/link";

export class CustomLinkNode extends LinkNode {
  constructor(url, attributes, key) {
    console.log("hello custom link");
    super(url, attributes, key);
  }

  static getType() {
    return "custom-link";
  }

  static clone(node) {
    return new CustomLinkNode(node.__url);
  }
}

// editor config
const initialConfig = {
  ...myConfig,
  nodes: [
    ...myConfig.nodes,
    // LinkNode,
    CustomLinkNode,
    {
      replace: LinkNode,
      with: (node) => {
        return new CustomLinkNode(node.getURL());
      }
    }
  ]
};

```

**Comment by kazukinagata** [comment]
**UPDATE:**

I think I found the cause by digging the code.
It seems to be the manner that all node instances are wrapped by [$applyNodeReplacement](https://github.com/facebook/lexical/blob/d30c88202e8d68e7ab79c9d2bf45366de0cea794/packages/lexical/src/LexicalUtils.ts#L1296) in `$create{***}Node` function, but LinkNode does not seem to follow this manner.

Wrapped:
- [LineBreakNode](https://github.com/facebook/lexical/blob/main/packages/lexical/src/nodes/LexicalLineBreakNode.ts#L91)
- [TextNode](https://github.com/facebook/lexical/blob/main/packages/lexical/src/nodes/LexicalTextNode.ts#L926)

Not:
- [LinkNode](https://github.com/facebook/lexical/blob/main/packages/lexical-link/src/index.ts#L248)

**Comment by abpbackup** [comment]
Hi. We have this problem as well since we use Electron and we need to open links using and Electron API so they are opened outside in a browser. The difference here with other node overrides is that this one needs to inherit a lot of functionalities and data from the editor but the HTML behavior needs to be changed. Here is our solution:

This is the custom link node which overrides the Lexical one.
```
import { EditorConfig } from 'lexical';
import { LinkNode } from '@lexical/link';
import { openExternalURL } from '../../../../services-electron/ipcUtil';
import { Platform } from '../../../../utils/platform';

export class CustomLinkNode extends LinkNode {
  static url: string;

  constructor(url: string) {
    super(url);
    this.url = url;
  }

  static getType() {
    return 'custom-link';
  }

  static clone(node) {
    return new CustomLinkNode(node.__key);
  }

  static onClick(url: string) {
    let fullUrl = url;
    if (!fullUrl.startsWith('http')) {
      fullUrl = 'https://' + fullUrl;
    }
    return Platform.isWeb ? window.open(fullUrl, '_blank') : openExternalURL(fullUrl);
  }

  static importJSON() {
    return null;
  }

  createDOM(config: EditorConfig) {
    const anchorElement = document.createElement('a');
    (anchorElement as HTMLAnchorElement).className = 'link';
    (anchorElement as HTMLAnchorElement).onclick = e => {
      e.preventDefault();
      CustomLinkNode.onClick(this.url);
    };

    return anchorElement;
  }
}
```

Then you need to tell Lexical to use your custom node.
```
...
<LexicalComposer
      initialConfig={
        {
          ...,
          nodes: [
            ...
            CustomLinkNode,
            {
              replace: LinkNode,
              with: node => {
                return new CustomLinkNode(node.getURL());
              },
            },
          ],
        } as InitialConfigType
      }
```

**Reply by abpbackup** [reply]
@kazukinagata ☝️ let me know if that helps
